modules/caddytls/tls.go

```diff
@@ -305,16 +324,44 @@ func (t *TLS) Stop() error {
 
 // Cleanup frees up resources allocated during Provision.
 func (t *TLS) Cleanup() error {
-	// stop the certificate cache
-	if t.certCache != nil {
-		t.certCache.Stop()
-	}
-
 	// stop the session ticket rotation goroutine
 	if t.SessionTickets != nil {
 		t.SessionTickets.stop()
 	}
 
+	// if a new TLS app was loaded, remove certificates from the cache that are no longer
+	// being managed or loaded by the new config; if there is no more TLS app running,
+	// then stop cert maintenance and let the cert cache be GC'ed
+	if nextTLS := caddy.ActiveContext().AppIfConfigured("tls"); nextTLS != nil {
```

Comment:
Oh I like this, that's nice 👍

Reply:
Yes! I was pleased I didn't have to keep pointers to app modules laying around. But I had to remove a lock within ActiveContext() to avoid deadlock.

I'm not 100% certain that lock was needed but I suspect it may have been. Race detector didn't catch anything yet. I will try to test more...